File: backend/app/schemas/ai_analysis.py
from typing import Any
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class AiAnalysisResponse(BaseModel):
    enabled: bool
    provider: str
    summary: str | None = None
    quality_score: int | None = None
    strengths: list[str] = Field(default_factory=list)
    issues: list[dict[str, Any]] = Field(default_factory=list)
    suggestions: list[str] = Field(default_factory=list)
    risk_level: str | None = None
    files_count: int | None = None
    message: str | None = None
    error: str | None = None


class AiAnalysisRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    repository_id: UUID
    status: str
    result_json: dict[str, Any] | None
    error_message: str | None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime

File: frontend/src/services/aiAnalysis.ts
import { API_URL, fetchWithTimeout, getAuthHeaders } from '@/services/api'

export type AiIssue = {
  severity: string
  category: string
  file: string
  description: string
  suggestion: string
}

export type AiAnalysisResponse = {
  enabled: boolean
  provider: string
  summary: string | null
  quality_score: number | null
  strengths: string[]
  issues: AiIssue[]
  suggestions: string[]
  risk_level: string | null
  files_count: number | null
  message: string | null
  error: string | null
}

async function parseApiError(response: Response): Promise<string> {
  try {
    const body = await response.json()
    if (typeof body.detail === 'string') {
      return body.detail
    }
  } catch {
    // Si el backend no devuelve JSON usamos un mensaje generico
  }

  return 'No se pudo completar el analisis IA'
}

const AI_ANALYSIS_TIMEOUT_MS = 120000

export async function analyzeRepositoryWithAi(repositoryId: string): Promise<AiAnalysisResponse> {
  const headers = await getAuthHeaders()

  if (!headers.Authorization) {
    throw new Error('Inicia sesion para analizar con IA')
  }

  const response = await fetchWithTimeout(
    `${API_URL}/repositories/${repositoryId}/ai-analysis`,
    {
      method: 'POST',
      headers,
    },
    AI_ANALYSIS_TIMEOUT_MS,
  )

  if (!response.ok) {
    throw new Error(await parseApiError(response))
  }

  return response.json()
}
